qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1986 @@
+# that they have been altered from the originals.
+
+r"""Use the pulse builder to write pulse programs with an imperative,
+assembly-like syntax.
+
+.. warning::
+    The pulse builder interface is still in active development and may have
+    breaking API changes without deprecation warnings in future releases until
+    otherwise indicated.
+
+To begin pulse programming we must first initialize our program builder
+context with :func:`build`, after which we can begin adding program
+statements. For example, below we write a simple program that :func:`play`\s
+a pulse:
+
+.. jupyter-execute::
+
+    from qiskit import execute
+    from qiskit import pulse
+
+    d0 = pulse.DriveChannel(0)
+
+    with pulse.build() as pulse_prog:
+        pulse.play(pulse.Constant(100, 1.0), d0)
+
+    pulse_prog.draw()
+
+The builder initializes a :class:`pulse.Schedule`, ``pulse_prog``
+and then begins to construct the program within the context. The output pulse
+schedule will survive after the context is exited and can be executed like a
+normal Qiskit program using ``qiskit.execute(pulse_prog, backend)``.
+
+Pulse programming has a simple imperative style. This leaves the programmer
+to worry about the raw experimental physics of pulse programming and not
+constructing cumbersome data structures.
+
+We can optionally pass a :class:`~qiskit.providers.BaseBackend` to
+:func:`build` to enable enhanced functionality. Below, we prepare a Bell state
+by automatically compiling the required pulses from their gate-level
+representations, while simultaneously applying a long decoupling pulse to a
+neighboring qubit. We terminate the experiment with a measurement to see which
+state we prepared. This program which mixes circuits and pulses will be
+automatically lowered to be run as a pulse program:
+
+.. jupyter-execute::
+
+    import math
+
+    from qiskit import pulse
+    from qiskit.test.mock import FakeOpenPulse3Q
+
+    # TODO: This example should use a real mock backend.
+    backend = FakeOpenPulse3Q()
+
+    d2 = pulse.DriveChannel(2)
+
+    with pulse.build(backend) as bell_prep:
+        pulse.u2(0, math.pi, 0)
+        pulse.cx(0, 1)
+
+    with pulse.build(backend) as decoupled_bell_prep_and_measure:
+        # We call our bell state preparation schedule constructed above.
+        with pulse.align_right():
+            pulse.call(bell_prep)
+            pulse.play(pulse.Constant(bell_prep.duration, 0.02), d2)
+            pulse.barrier(0, 1, 2)
+            registers = pulse.measure_all()
+
+    decoupled_bell_prep_and_measure.draw()
+
+With the pulse builder we are able to blend programming on qubits and channels.
+While the pulse IR of Qiskit is based on instructions that operate on
+channels, the pulse builder automatically handles the mapping from qubits to
+channels for you.
+
+In the example below we demonstrate some more features of the pulse builder:
+
+.. jupyter-execute::
```

Comment:
This whole block of code should be broken up into many blocks. This is too intimidating for the users we want to help. It's probably best if you make a new issue to do this with the rest of the builder docs/tutorials, rather than fixing it in this PR.